massa-executed-ops/src/executed_ops.rs

```diff
@@ -0,0 +1,345 @@
+    streaming_step::StreamingStep,
+    wrapped::Id,
+};
+use massa_serialization::{
+    Deserializer, SerializeError, Serializer, U64VarIntDeserializer, U64VarIntSerializer,
+};
+use nom::{
+    error::{context, ContextError, ParseError},
+    multi::length_count,
+    sequence::tuple,
+    IResult, Parser,
+};
+use std::{
+    collections::VecDeque,
+    ops::Bound::{Excluded, Included, Unbounded},
+};
+
+const EXECUTED_OPS_INITIAL_BYTES: &[u8; 32] = &[0; HASH_SIZE_BYTES];
+
+/// A structure to list and prune previously executed operations
+#[derive(Debug, Clone)]
+pub struct ExecutedOps {
+    /// Exectued operations configuration
+    config: ExecutedOpsConfig,
+    /// Executed operations deque associated with a Slot for better pruning complexity
+    pub ops_deque: VecDeque<(Slot, PreHashSet<OperationId>)>,
+    /// Executed operations only for better insertion complexity
+    ops: PreHashSet<OperationId>,
+    /// Accumulated hash of the executed operations
+    pub hash: Hash,
+}
+
+impl ExecutedOps {
+    /// Creates a new `ExecutedOps`
+    pub fn new(config: ExecutedOpsConfig) -> Self {
+        Self {
+            config,
+            ops_deque: VecDeque::new(),
+            ops: PreHashSet::default(),
+            hash: Hash::from_bytes(EXECUTED_OPS_INITIAL_BYTES),
+        }
+    }
+
+    /// Returns the number of executed operations
+    pub fn len(&self) -> usize {
+        self.ops.len()
+    }
+
+    /// Check executed ops emptiness
+    pub fn is_empty(&self) -> bool {
+        self.ops.is_empty()
+    }
+
+    /// Internal function used to insert the values of an operation id iter and update the object hash
+    fn extend_and_compute_hash<'a, I>(&mut self, values: I)
+    where
+        I: Iterator<Item = &'a OperationId>,
+    {
+        for op_id in values {
+            if self.ops.insert(*op_id) {
+                self.hash ^= *op_id.get_hash();
+            }
+        }
+    }
+
+    /// Apply speculative operations changes to the final executed operations state
+    pub fn apply_changes(&mut self, changes: ExecutedOpsChanges, slot: Slot) {
+        self.extend_and_compute_hash(changes.iter());
+        match self.ops_deque.back_mut() {
+            Some((last_slot, ids)) if *last_slot == slot => {
+                ids.extend(changes);
+            }
+            Some((last_slot, _)) => match last_slot.get_next_slot(self.config.thread_count) {
+                Ok(next_to_last_slot) if next_to_last_slot == slot => {
+                    self.ops_deque.push_back((next_to_last_slot, changes));
+                }
+                Ok(next_to_last_slot) => {
+                    panic!("executed ops associated slot must be sequential, expected {:?} but received {:?}", next_to_last_slot, slot)
+                }
+                _ => panic!("get_next_slot overflow"),
+            },
+            None => {
+                self.ops_deque.push_back((slot, changes));
+            }
+        }
+        self.prune(slot);
```

Comment:
I don't understand why we prune here. Reading the code of the pruning I understand that it remove all executed ops before this slot, maybe I misunderstood but otherwise that means that we remove all ops that are before this slot which is a problem for example in bootstrap no ?

Reply:
I just noticed a bad behaviour in the executed ops refactor thanks to your comment, it is a little tricky to fix I will explain the process once it is done